proposals/NNNN-objc-implementation.md

```diff
@@ -0,0 +1,323 @@
+
+`@implementation` cannot create declarations that aren't supported by `@objc`. The most notable limitations include:
+
+* Free functions, global variables, cases of `NS_TYPED_ENUM` typedefs, and other non-member Objective-C declarations.
+* Global declarations imported as members of a type using `NS_SWIFT_NAME`'s import-as-member capabilities.
+* Factory convenience initializers (those implemented as class methods, like `+[NSString stringWithCharacters:length:]`).
+* `__attribute__((objc_direct))` methods and `@property (direct)` properties.
+* Members with nonstandard memory management behavior, even if it is correctly annotated.
+* Members which deviate from the Objective-C error convention in certain subtle ways, such as by having the `NSError**` parameter in the wrong place.
+
+`@objc @implementation` heavily piggybacks on `@objc`'s code emission, so in most cases, the best approach to expanding `@implementation`'s support would be to extend `@objc` to support the feature and then make sure `@objc @implementation` supports it too.
+
+### `@implementation` for plain C declarations
+
+Many of the capabilities mentioned as future directions for `@objc` would also be useful for plain-C clients, including those on non-Darwin platforms. Once again, the best approach here would probably be to stabilize and extend something like `@_cdecl` to support creating these with a generated header, and then make sure `@implementation` supports this attribute too.
+
+The compiler currently has experimental support for `@_cdecl @implementation` for global functions; it's behind a separate experimental feature flag because it's not part of this proposal.
+
+### `@implementation` for C++ declarations
+
+One could similarly imagine a C++ version of this feature:
+
+```cpp
+// C++ header file
+
+class CppClass {
+private:
+    int myStorage = 0;
+public:
+    int someMethod() { ... }
+    int swiftMethod();
+};
+```
+
+```swift
+// Swift implementation file
+
+@_expose(Cxx) @implementation extension CppClass {
+    func swiftMethod() -> Int32 { return self.myStorage }
+}
+```
+
+This would be tricker than Objective-C interop because for Objective-C interop, Swift already generates machine code thunks directly in the binary, whereas for C++ interop, Swift generates C++ source code thunks in a generated header. Swift could either compile this generated code internally, or it could emit it to a file and expect the build system to build and link it.
+
+We believe that there wouldn't be a problem with sharing the `@implementation` attribute with this feature because `@implementation` is always paired with a language-specific attribute
+
+### Supporting lightweight generics
+
+Objective-C's "lightweight generics" behave in rather janky ways in Swift extensions, so we have banned them in this initial document. If there’s demand for implementing Objective-C generic classes in Swift, we may want to extend this feature to support them.
```

Comment:
"rather janky" feels.. a bit flippant for an evolution proposal :)

Reply:
```suggestion
Classes using Objective-C lightweight generics have type-erased generic parameters; this imposes a lot of tricky limitations on Swift extensions of these classes. Since very few classes use lightweight generics, we have chosen to ban the combination for now. If there turns out to be a lot of demand for implementing Objective-C generic classes in Swift, we can lift the ban after we've figured out how to make the combination more usable.
```